### PR TITLE
Fix camera not working on the second time opening it

### PR DESCRIPTION
**Pull request by QichenZhu**
### Fixed Issues

$ https://github.com/Expensify/App/issues/51337
PROPOSAL: https://github.com/Expensify/App/issues/51337#issuecomment-2496966988

### Tests

1. Open New Expensify in mobile Chrome.
2. Tap FAB > Create Expense > Scan.
3. If camera permission isn't granted yet, tap Continue and Allow.
4. Take a picture.
5. Tap the back arrow (<) in the top left corner to return to the Scan page.
6. Verify the camera works.

- [x] Verify that no errors appear in the JS console

### Offline tests

1. Open New Expensify in mobile Chrome.
2. Tap FAB > Create Expense > Scan.
3. If camera permission isn't granted yet, tap Continue and Allow.
4. Take a picture.
5. Tap the back arrow (<) in the top left corner to return to the Scan page.
6. Verify the camera works.

### QA Steps

1. Open New Expensify in mobile Chrome.
2. Tap FAB > Create Expense > Scan.
3. If camera permission isn't granted yet, tap Continue and Allow.
4. Take a picture.
5. Tap the back arrow (<) in the top left corner to return to the Scan page.
6. Verify the camera works.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I used [JaimeGPT](https://chatgpt.com/g/g-2dgOQl5VM-english-to-spanish-translator-aka-jaimegpt) to get English > Spanish translation. I then posted it in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.ts or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I added [unit tests](https://github.com/Expensify/App/blob/main/tests/README.md) for any new feature or bug fix in this PR to help automatically prevent regressions in this user flow.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/00f3c0d9-4ae1-4da7-b9a7-777cecb08f06

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/a07f545a-f378-40a1-a70a-8ee6c8c63abb

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/6bebba37-64ae-4465-9742-8a40b12e9bc5

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/6d42324e-42cf-43ef-b794-c6be8032c07a

https://github.com/user-attachments/assets/05f20fa2-4f9d-4e0d-a792-45c389623ba0

</details>

<details>
<summary>MacOS: Chrome / Safari (Not Applicable)</summary>

<p></p>

Test steps are not applicable on this platform.

mac-web

</details>

<details>
<summary>MacOS: Desktop (Not Applicable)</summary>

<p></p>

Test steps are not applicable on this platform.

mac-desktop

</details>

### CODE REVIEW

**Comment by QichenZhu** [comment]
@ishpaul777 This PR is almost ready, but two issues remain:

1. The camera view is sometimes cut off as mentioned before. This is a suspected commit: https://github.com/Expensify/App/pull/28411/commits/98519f0630ca70d12892589c6761e1af14074a38.

2. While repeating the test steps, the camera randomly stops working on some devices (see 00:43 in the video). Still looking into it.

https://github.com/user-attachments/assets/a6785414-7bef-46b0-a8e9-65acf6c14505

**Comment by QichenZhu** [comment]
@ishpaul777 I need your guidance on whether the two mentioned issues are blockers for this PR and the next steps.

**Comment by QichenZhu** [comment]
@ishpaul777 The two related issues seem resolved.

**Comment by QichenZhu** [comment]
@ishpaul777 Sure! Actually, there were 3 (maybe 4) different problems.

#### 1. Main issue #51337
`requestCameraPermission()` used `videoConstraints` but didn’t include it in the dependency array. Resolved by removing the use of `videoConstraints` from `requestCameraPermission()` and recalculating it every time the camera starts.

#### 2. Random camera crash on some old iOS versions
Sometimes the camera started normally but stopped working within one second. Reproducible in Safari and Chrome on iOS 17.5 and 17.7.2 but seems resolved on iOS 18. I suspect the camera view resized on startup, causing iOS to fail rendering it. Fixed by setting `display` to `none` during this process.

#### 3. Cropped camera view
I think this was just another form of the same issue as above on different iOS versions.

#### 4. Random camera init failure in Chrome
Safari and Chrome behaved differently because `Browser.isMobileSafari()` only matches Safari. Resolved by changing it to `Browser.isMobileWebkit()`, which matches both Safari and Chrome, so they initialize in the same way.

https://github.com/Expensify/App/blob/3cc88f5a83a979c5e588e01065083a3d714e69da/src/pages/iou/request/step/IOURequestStepScan/index.tsx#L132

**Comment by QichenZhu** [comment]
> Bug: App crashes on android when navigating back after scanning receipt

@ishpaul777 This PR only changes web files and types, so it's unlikely to cause this. I just tested merging this PR into `main` and it works fine on Android. Do you need me to merge `main` into this PR?

**Comment by QichenZhu** [comment]
Thanks! Checklist updated.

**Comment by QichenZhu** [comment]
> Can you please add unit tests ? https://expensify.enterprise.slack.com/archives/C01GTK53T8Q/p1733148961659549

I suggest not adding unit tests because: 1. The main issue can easily be avoided by eslint if we hadn't suppressed the warning. 2. The other related issue requires running on a specific iOS version on a physical device to reproduce, which is not practical to mock.

**Review by ishpaul777 — APPROVED** [review]

**Review by techievivek — APPROVED** [review]

**Comment by techievivek** [comment]
I have checked off `For any bug fix or new feature in this PR, I verified that sufficient [unit tests](https://github.com/Expensify/App/blob/main/tests/README.md) are included to prevent regressions in this flow.` based on above points so we can move forward with merge.

File: src/pages/iou/request/step/IOURequestStepScan/NavigationAwareCamera/WebCamera.tsx [hunk 1]
```diff
@@ -1,23 +1,29 @@
...
+    const styles = useThemeStyles();
 
     if (!shouldShowCamera) {
         return null;
     }
+
     return (
-        <View>
+        // Hide the camera during initialization to prevent random failures on some iOS versions.
+        <View style={isInitialized ? styles.dFlex : styles.dNone}>
```
Comment (dominictb) [inline src/pages/iou/request/step/IOURequestStepScan/NavigationAwareCamera/WebCamera.tsx:22]: Missing `flex: 1` caused this issue https://github.com/Expensify/App/issues/64049

Comment (QichenZhu) [reply src/pages/iou/request/step/IOURequestStepScan/NavigationAwareCamera/WebCamera.tsx:22]: Thanks for the comment! But this PR didn't cause that. We already used `<View>` without `flex: 1` on purpose before to solve [an issue](https://github.com/Expensify/App/pull/28411#issuecomment-1756187724) (commit: https://github.com/Expensify/App/pull/28411/commits/98519f0630ca70d12892589c6761e1af14074a38) and `dFlex` was the default. This PR only added `dNone`.

**Pull request merged**